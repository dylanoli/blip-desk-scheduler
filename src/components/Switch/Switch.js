import React, { useState, useEffect } from 'react';
// eslint-disable-next-line import/no-unresolved
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './styles.module.scss';

const Switch = ({
    name,
    label,
    labelPosition = 'left',
    size = 'standard',
    disabled = false,
    checked = false,
    rounded = true,
    onChange = () => {}
}) => {
    const [isChecked, setIsChecked] = useState(false);

    useEffect(() => {
        setIsChecked(checked);
    }, [checked]);

    const handleChange = () => {
        onChange(!checked);
    };

    return (
        <label
            className={classNames(
                'relative flex w-100 items-center',
                styles.switch,
                {
                    [styles.short]: size === 'short',
                    [styles.tall]: size === 'tall',
                    [styles.gap]: !!label
                }
            )}
        >
            {!!label && labelPosition === 'left' && (
                <span className="b f6 bp-c-extend-grays-street">{label}</span>
            )}
            <input
                data-testid="switch-input"
                type="checkbox"
                name={name}
                checked={isChecked}
                disabled={disabled}
                onChange={handleChange}
            />
            <span
                className={classNames(
                    'relative pointer top-0 left-0 right-0 bottom-0 bp-bg-neutral-medium-silver',
                    styles.slider,
                    {
                        [styles.round]: !!rounded
                    }
                )}
            ></span>
            {!!label && labelPosition === 'right' && (
                <span className="b f6 bp-c-extend-grays-street">{label}</span>
            )}
        </label>
    );
};

Switch.propTypes = {
    name: PropTypes.string,
    label: PropTypes.string,
    labelPosition: PropTypes.string,
    size: PropTypes.string,
    checked: PropTypes.bool,
    disabled: PropTypes.bool,
    rounded: PropTypes.bool,
    onChange: PropTypes.func
};

export default Switch;
