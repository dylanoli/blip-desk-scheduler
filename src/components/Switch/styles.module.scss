.switch {
    input {
        opacity: 0;
        width: 0;
        height: 0;
    }

    .slider {
        width: 42px;
        height: 24px;

        transition: 0.4s;

        &::before {
            position: absolute;
            content: '';
            background-color: white;
            transition: 0.4s;
            height: 18px;
            width: 18px;
            left: 2.75px;
            bottom: 3.25px;
        }

        &.round {
            border-radius: 34px;

            &::before {
                border-radius: 50%;
            }
        }
    }

    input:checked + .slider {
        background-color: #3f7de8;
    }

    input:disabled + .slider {
        background-color: #d2dfe6;
        cursor: not-allowed;
    }

    input:checked + .slider:before {
        transform: translateX(18px);
    }

    &.short {
        .slider {
            width: 32px;
            height: 18px;

            &::before {
                height: 13.5px;
                width: 13.5px;
                left: 2.25px;
                bottom: 2.25px;
            }
        }

        input:checked + .slider:before {
            transform: translateX(13.75px);
        }
    }

    &.tall {
        .slider {
            width: 56px;
            height: 32px;

            &::before {
                width: 24px;
                height: 24px;
                left: 4px;
                bottom: 4px;
            }
        }

        input:checked + .slider:before {
            transform: translateX(24px);
        }
    }

    &.gap {
        gap: 0.25rem;
    }
}
