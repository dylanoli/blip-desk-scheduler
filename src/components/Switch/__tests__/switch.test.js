import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';

import Switch from '../Switch';

describe('Switch component', () => {
    const handleChange = jest.fn();

    it('should render correctly', () => {
        const { container } = render(
            <Switch name="test" checked={false} onChange={handleChange} />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
