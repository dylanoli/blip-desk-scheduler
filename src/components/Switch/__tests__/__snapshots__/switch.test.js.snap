// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component should render correctly 1`] = `
<label
  class="relative flex w-100 items-center switch"
>
  <input
    data-testid="switch-input"
    name="test"
    type="checkbox"
  />
  <span
    class="relative pointer top-0 left-0 right-0 bottom-0 bp-bg-neutral-medium-silver slider round"
  />
</label>
`;
